velox/common/tests/AsyncSourceTest.cpp

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/common/base/AsyncSource.h"
+#include <fmt/format.h>
+#include <folly/Random.h>
+#include <folly/Synchronized.h>
+#include <gtest/gtest.h>
+#include <thread>
+
+using namespace facebook::velox;
+
+// A sample class to be constructed via AsyncSource.
+struct Gizmo {
+  Gizmo(int32_t _id) : id(_id) {}
+
+  int32_t id;
+};
+
+TEST(AsyncSourceTest, basic) {
```

Comment:
I would expect the "basic" test to test the basic API, like construct a single async source, check `isPending()` and `hasValue()`, then call `prepare()`, check the method return the expected results again, then move the results out, etc.

Then having the test below as a "multi-thread" test that checks for concurrent behavior.